File: scala/mergesort.scala
def mergesort[T](xs: Seq[T])(implicit ord: Ordering[T]): Seq[T] = {
  def merge(left: Seq[T], right: Seq[T]): Seq[T] = {
    (left, right) match {
      case (Seq(), rs) => rs
      case (ls, Seq()) => ls
      case (Seq(l, lr @ _*), Seq(r, rr @ _*)) =>
        if (ord.lt(l, r))
          l +: merge(Seq(lr: _*), right)
        else
          r +: merge(left, Seq(rr: _*))
    }
  }

  val n = xs.length / 2
  if (n == 0)
    xs
  else {
    val (left, right) = xs.splitAt(n)
    merge(mergesort(left), mergesort(right))
  }
}

println(mergesort(Seq(10, 9, 8, 4, 5, 6, 7, 3, 2, 1)).mkString(" "))

File: nim/mergesort.nim
proc merge(left, right: var seq[int]): seq[int] =
  var list: seq[int]
  list = @[]
  while len(left) > 0 and len(right) > 0:
    if left[0] <= right[0]:
      list.add(left[0])
      left = left[1..^1]
    else:
      list.add(right[0])
      right = right[1..^1]
  while len(left) > 0:
    list.add(left[0])
    left = left[1..^1]
  while len(right) > 0:
    list.add(right[0])
    right = right[1..^1]
  return list

proc mergesort*(a: seq[int]): seq[int] =
  if a.len <= 1:
    return a
  var left : seq[int]
  var right: seq[int]
  for k, v in a:
    if k < int(len(a)/2):
      left.add(v)
    else:
      right.add(v)
  left = mergesort(left)
  right = mergesort(right)
  return merge(left, right)

var numbers = @[10, 9, 8, 4, 5, 6, 7, 3, 2, 1]
echo mergesort(numbers)
